srml/support/procedural/src/storage/transformation.rs

```diff
@@ -210,34 +210,42 @@ fn decl_store_extra_genesis(
 				} else {
 					quote!( #d )
 				}).unwrap_or_else(|| quote!( Default::default() ));
-					config_field_default.extend(quote!( #ident: #fielddefault, ));
-
+			config_field_default.extend(quote!( #ident: #fielddefault, ));
 		} else {
 			opt_build = build.as_ref().map(|b| &b.expr.content).map(|b| quote!( #b ));
 		}
 
 		let typ = type_infos.typ;
 		if let Some(builder) = opt_build {
 			is_trait_needed = true;
-			if type_infos.is_simple {
-				builders.extend(quote!{{
-					use #scrate::codec::Encode;
-					let v = (#builder)(&self);
-					r.insert(Self::hash(
-						<#name<#traitinstance> as #scrate::storage::generator::StorageValue<#typ>>::key()
-						).to_vec(), v.encode());
-				}});
-			} else {
-				let kty = type_infos.map_key.clone().expect("is not simple; qed");
-				builders.extend(quote!{{
-					use #scrate::codec::Encode;
-					let data = (#builder)(&self);
-					for (k, v) in data.into_iter() {
-						let key = <#name<#traitinstance> as #scrate::storage::generator::StorageMap<#kty, #typ>>::key_for(&k);
-						r.insert(Self::hash(&key[..]).to_vec(), v.encode());
-					}
-				}});
-			}
+			builders.extend(match type_infos.kind {
+				DeclStorageTypeInfosKind::Simple => {
+					quote!{{
+						use #scrate::rstd::{cell::RefCell, marker::PhantomData};
+						use #scrate::codec::{Encode, Decode};
+
+						let storage = (RefCell::new(&mut r), PhantomData::<Self>::default());
+						let v = (#builder)(&self);
+						let v = Encode::using_encoded(&v, |mut v| Decode::decode(&mut v))
+							.expect("Genesis parameters encoding does not match the expected type.");
```

Comment:
Can we add the name of parameter here as well? So it is easier to debug in a case of a failure.